src/lib/udev/google_nvme_id

```diff
@@ -0,0 +1,305 @@
+function get_namespace_device_name() {
+  local nvme_json
+  nvme_json="$("${nvme_cli_bin}" id-ns -b "${1}" | xxd -p -seek 384 -l 256 | xxd -p -r)"
+  if [[ $? != 0 ]]; then
+    return 1
+  fi
+
+  if [[ -z "${nvme_json}" ]]; then
+    err "NVMe Vendor Extension disk information not present"
+    return 1
+  fi
+
+  local device_name
+  device_name="$(echo "${nvme_json}" | sed -e 's/.*"device_name":"\([a-zA-Z0-9_-]\+\)".*/\1/g')"
+
+  # Error if our device name is empty
+  if [[ -z "${device_name}" ]]; then
+    err "Empty name"
+    return 1
+  fi
+
+  echo "${device_name}"
+  return 0
+}
+
+#######################################
+# Retrieves the nsid for an NVMe namespace
+# Globals:
+#   None
+# Arguments:
+#   The path to the nvme namespace (/dev/nvme0n*)
+# Outputs:
+#   The namespace number/id
+# Returns:
+#   0 if the namespace id could be retrieved, 1 otherwise
+#######################################
+function get_namespace_number() {
+  local dev_path="${1}"
+  local namespace_number
+  if [[ "${dev_path}" =~ $NAMESPACE_NUMBER_REGEX ]]; then
+    namespace_number="${BASH_REMATCH[1]}"
+  else
+    return 1
+  fi
+
+  echo "${namespace_number}"
+  return 0
+}
+
+#######################################
+# Retrieves the partition number for a device path if it exists
+# Globals:
+#   None
+# Arguments:
+#   The path to the device partition (/dev/nvme0n*p*)
+# Outputs:
+#   The value after 'p' in the device path, or an empty string if the path has
+#   no partition.
+#######################################
+function get_partition_number() {
+  local dev_path="${1}"
+  local partition_number
+  if [[ "${dev_path}" =~ $PARTITION_NUMBER_REGEX ]]; then
+    partition_number="${BASH_REMATCH[1]}"
+    echo "${partition_number}"
+  else
+    echo ''
+  fi
+  return 0
+}
+
+#######################################
+# Generates a symlink for a PD-NVMe device using the metadata's disk name.
+# Primarily used for testing but can be used if the script is directly invoked.
+# Globals:
+#   Uses ID_SERIAL_SHORT (can be populated by identify_pd_disk)
+# Arguments:
+#   The device path for the disk
+#######################################
+function gen_symlink() {
+  local dev_path="${1}"
+  local partition_number="$(get_partition_number "${dev_path}")"
+
+  if [[ -n "${partition_number}" ]]; then
+    ln -s "${dev_path}" /dev/disk/by-id/google-persistent-disk-"${ID_SERIAL_SHORT}"-part"${partition_number}" > /dev/null 2>&1
+  else
+    ln -s "${dev_path}" /dev/disk/by-id/google-persistent-disk-"${ID_SERIAL_SHORT}" > /dev/null 2>&1
+  fi
+
+  return 0
+}
+
+#######################################
+# Validates that a given path corresponds to a valid NVMe device
+# Globals:
+#   Uses nvme_cli_bin
+# Arguments:
+#   The device path for the disk
+# Returns:
+#   The return value of the call to the nvme id-ctrl command
+#######################################
+function validate_nvme_device() {
+  "${nvme_cli_bin}" id-ctrl "${1}" &>/dev/null
```

Comment:
this single line can be used in place of this function

Reply:
Fixed in most recent commit.